Util: Allow scheduler to be mocked

amitiuttarwar opened the pull request:
This PR is to support functional tests by allowing the scheduler to be mocked via the RPC. 

It adds a `MockForward` method to the scheduler class that iterates through the task queue and reschedules them to be `delta_seconds` sooner. 

This is currently used to support functional testing of the "unbroadcast" set tracking in #18038. If this patch is accepted, it would also be useful to simplify the code in #16698.

amitiuttarwar committed:
[util] allow scheduler to be mocked

Add MockForward method to the scheduler that mimics going into the future by rescheduling all items on the taskQueue to be sooner.

amitiuttarwar committed:
[test] unit test for new MockForward scheduler method

amitiuttarwar commented:
Some relevant info about the analysis of the cost (maintenance burden) vs benefit (supporting a better tested codebase) of these changes 
pros: 
- these changes add a function to the scheduler with unit test coverage
- the function would only be used when running regtest
- intent is for this to support easier functional testing for other features 
- currently, two explicit use cases 
cons: 
- all code adds maintenance burden 
- c++ code specifically to support testing should only be added when necessary 
- the scheduler mock is not required for functionally testing the current use cases. workarounds are possible 

explanation of the two use cases & workarounds.
1. [INSERT PR] uses the mock to trigger the next attempt at initial broadcast, which is scheduled for every 10 minutes. Alternatively, the `REATTEMPT_BROADCAST_INTERVAL` constant that indicates scheduler frequency could be defined as a chain param and made more frequent for regtest (thanks @MarcoFalke for the suggestion).
2. In PR #16698, specifically commit baf7de22108eb9081ddc700067b72409efbc382e, the mock would allow deleting some of the c++ code (but still be testable.) In the patch right now, the scheduler periodically calls `MaybeResendWalletTxs` which calls `ResubmitWalletTransactionsToMempool` which uses `nNextResend` to track time. `nNextResend` implements a mockable time so can be tested from the functional tests. If the scheduler mock is supported, the scheduler could directly invoke the resubmit function & drop the member var to maintain time tracking.

JeremyRubin commented:
utACK 58c4c19.

This seems like a generally useful test harness, and the code looks basically correct to me. I think the pointer to the scheduler is a little crass, but it isn't a bad option considering the scheduler is created as a static in init.cpp (someone can clean it up later if they make the schedule a non static).

The time delta technically could underflow, but I don't think that's the base time being subtracted from is a time point (and we're far from the unix epoch time). If this weren't just testing harness, it might be worth it to check for underflow, but since we'd basically be having to actively try to shoot ourselves in the foot during a test to get an underflow this usage is fine. 

Great work!

fanquake commented:
Concept ACK. 

Looks like there's an issue with the new `mockforward` unit test, as most of the [Travis instances](https://travis-ci.org/bitcoin/bitcoin/builds/644207906) are failing:
```bash
test/scheduler_tests.cpp(158): Entering test case "mockforward"
test/scheduler_tests.cpp(188): error: in "scheduler_tests/mockforward": check num_tasks == 1 has failed [3 != 1]
test/scheduler_tests.cpp(191): error: in "scheduler_tests/mockforward": check counter == 2 has failed [0 != 2]
test/scheduler_tests.cpp(197): error: in "scheduler_tests/mockforward": check delta > 2*60 && delta < 3*60 has failed
test/scheduler_tests.cpp(158): Leaving test case "mockforward"; testing time: 6321us
test/scheduler_tests.cpp(13): Leaving test suite "scheduler_tests"; testing time: 146946us
```

MarcoFalke commented:
No objection, but some observations:

* In light of the test failures on some platforms, we obviously need to make sure this is doing the right thing. Timestamp handling isn't obviously trivial especially in combination with condition variables. This seems true given the travis failures on some machines, but not others and also in light of the numerous scheduler crashed that no one can explain: #16307, #16027, #14200, ...

* "Mocking" code so that tests run the mocked code and production runs a different code might defeat the whole point of having the tests in the first place. Tests (especially functional tests) are supposed to test the production code (paths) as closely as possible.

So my personal preference would be to go with some minimally invasive and simple approach like the chainparam. But if everyone else likes this mock approach because it could be used for other testing needs as well, then I won't object having this merged.

kristapsk commented on src/rpc/misc.cpp:
```diff
@@ -366,6 +369,27 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+        }.Check(request);
+
+    if (!Params().MineBlocksOnDemand())
+        throw std::runtime_error("mockscheduler for regression testing (-regtest mode) only");
```
nit: "mockscheduler **is** for regression testing (-regtest mode) only"?

promag commented on src/node/context.h:
```diff
@@ -11,6 +11,7 @@
 class BanMan;
 class CConnman;
 class CTxMemPool;
+class CScheduler;
```
58c4c1943e082060f38054ed4738478635ab9c31

nit, keep sorted.

promag commented on src/rpc/misc.cpp:
```diff
@@ -5,15 +5,18 @@
 #include <rpc/server.h>
 #include <rpc/util.h>
 #include <script/descriptor.h>
+#include <scheduler.h>
```
58c4c1943e082060f38054ed4738478635ab9c31

nit, keep sorted.

promag commented on src/rpc/misc.cpp:
```diff
@@ -5,15 +5,18 @@
 #include <util/strencodings.h>
 #include <util/system.h>
 #include <util/validation.h>
+#include <init.h>
```
58c4c1943e082060f38054ed4738478635ab9c31

Why is this included?

amitiuttarwar replied:
good catch. was left over from a previous iteration

promag commented on src/rpc/misc.cpp:
```diff
@@ -366,6 +369,27 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+
+    RPCTypeCheck(request.params, {UniValue::VNUM});
+
+    g_rpc_node->scheduler->MockForward(boost::chrono::seconds(request.params[0].get_int64()));
```
58c4c1943e082060f38054ed4738478635ab9c31

Maybe check `delta_time > 0`?

promag reviewed (commented):
Concept ACK.

Some minor comments otherwise change LGTM.

promag commented on src/test/scheduler_tests.cpp:
```diff
@@ -155,4 +155,46 @@ BOOST_AUTO_TEST_CASE(singlethreadedscheduler_ordered)
+
+    // finish up
+    MicroSleep(600);
+    scheduler.stop(false);
```
ca57312e5b0dc9d28766369b963395f1690e0b4f

Maybe change this to:
```
scheduler.scheduleFromNow([&]() { scheduler.stop(false); }, 1);
```
and remove the sleep above.

amitiuttarwar replied:
clever. thank you! 

lets see if travis likes it 🤞🏽

amitiuttarwar replied:
yay! looking good. only travis failure seemed unrelated.

amitiuttarwar commented:
I addressed review comments & tests are now passing CI. Added checks in RPC function & scheduler function to ensure reasonable values / no underflow.

This PR has received 3 ACKs - utACK from @JeremyRubin & concept ACKs from @promag @fanquake. Would appreciate re-review 🙏🏽

JeremyRubin commented:
re-utACK ecb0ad5

promag commented on src/rpc/misc.cpp:
```diff
@@ -570,6 +600,7 @@ static const CRPCCommand commands[] =
 
     /* Not shown in help */
     { "hidden",             "setmocktime",            &setmocktime,            {"timestamp"}},
+    { "hidden",             "mockscheduler",          &mockscheduler,          {"delta_time"}},
```
Follow up, would be cool to skip this when !regtest and in `mockscheduler` would be `assert(Params().MineBlocksOnDemand())`.

JeremyRubin replied:
I think it's reasonable for @amitiuttarwar to just copy exactly what setmocktime is doing in this regard.

ajtowns replied:
Huh? `static const commands[]=` will get done prior to working out what chain is in use, so you'd have to do a much more serious rework to remove it from the list of rpcs entirely.

sipa replied:
It's a hidden RPC so I don't think this matters much.

promag replied:
@ajtowns I was thinking that `CRPCTable::execute` could take into account some `CRPCCommand` flag and then it would result in `RPC_METHOD_NOT_FOUND`.

promag reviewed (commented):
@amitiuttarwar sure, will review later.

promag commented on src/scheduler.cpp:
```diff
@@ -114,6 +114,27 @@ void CScheduler::scheduleFromNow(CScheduler::Function f, int64_t deltaMilliSecon
+        std::multimap<boost::chrono::system_clock::time_point, Function> temp_queue;
+
+        for (const auto& element : taskQueue) {
+            boost::chrono::system_clock::time_point new_time = std::max(boost::chrono::system_clock::now() + delta_seconds, element.first) - delta_seconds;
```
What's the problem of doing
```cpp
temp_queue.emplace_hint(temp_queue.cend(), element.first - delta_seconds, element.second);
```
Looks fine to me since `wait_until` returns when system time is later than that absolute time.

Didn't check the case for boost < 1.50.

JeremyRubin replied:
element.first - delta_seconds may underflow, so std::max prevents underflow.

promag replied:
Underflow how? You mean be less than current time?

amitiuttarwar replied:
if `element.first - delta_seconds < now` everything will process fine (as in the previous version you reviewed)

but if `delta_seconds > element.first` then there could be an underflow. seems highly unlikely because `element.first` is a time point, but doesn't hurt to have this guarantee

sipa replied:
This looks equivalent to

```
std::max(boost::chrono::system_clock::now(), element.first - delta_seconds);
```

Also perhaps you only want to invoke `now()` once and cache it rather than querying it for every queue element.

JeremyRubin replied:
@sipa that misses the point above. if delta_seconds > element.first, it underflows and then it can be greater than now().

I guess equivalents, if delta_seconds is too large, now() + delta_seconds may overflow... but that seems less likely because we don't know if this system_clock::now() which feeds into the queue is set to start at time 0 on startup.


Note also that caching now(), were emplace hint not used, would mean that the tmp queue would potentially be unstable ordered w.r.t. jobs with the same timeout.

sipa replied:
> if delta_seconds > element.first, it underflows and then it can be greater than now().

According to https://en.cppreference.com/w/cpp/chrono/duration, the representation type for the standard clock values is signed, so underflow should not be any issue (or at least not more than we need to worry about the clock overflowing in the first place).

> Note also that caching now(), were emplace hint not used, would mean that the tmp queue would potentially be unstable ordered w.r.t. jobs with the same timeout.

If that's a concern, it already is. There is no guarantee that subsequent calls to `now()` will be strictly increasing.

amitiuttarwar replied:
caching `now()` seems good, I will update. 

RE: underflow / overflow- 
the intent is to ensure `delta_seconds` is reasonable. I'm thinking of asserting the value is between 0 & 3600 from both this `MockForward` method & the `mockscheduler` rpc method. 
and reverting to the original `temp_queue.emplace_hint(temp_queue.cend(), element.first - delta_seconds, element.second)`

JeremyRubin replied:
Ah -- note that it's boost and not std... I think that the seconds duration is guaranteed to be signed 64, but I couldn't find anything on system_clock's duration. Good catch on the non steady adjustable clock of system_clock, I had missed that. Seems like a general thing that our scheduler should be on a monotonic hardware clock rather than something that adjusting the system date could adjust (which would be another way to do this patch -- just fast forward the system time).

sipa replied:
Even with a monotonic clock there is no guarantee that subsequent requests to `now()` are *strictly* increasing (they'll be *monotonically* increasing, but that's not enough to guarantee a stable sorting order absent hints).

amitiuttarwar replied:
reverted to original (`element.first - delta_seconds`), but added checks in `MockForward` and `mockscheduler` that `delta_seconds` is between 0 and 3600. 

which means we no longer invoke `now()` =P 

lmk if there are any further concerns

promag replied:
lgtm

promag commented on src/test/scheduler_tests.cpp:
```diff
@@ -155,4 +155,45 @@ BOOST_AUTO_TEST_CASE(singlethreadedscheduler_ordered)
+    CScheduler scheduler;
+
+    int counter{0};
+    CScheduler::Function dummy = [&counter](){counter++;};
```
nit, you can drop `()`.

JeremyRubin replied:
I think this is a style preference (plus it's in tests) so it's whatever.

For whatever reason, I prefer this form when it's a function that should be called multiple times (as dummy is) and the no () form when it's more akin to a std::bind for a call-once thing.

amitiuttarwar replied:
updated

promag commented on src/test/scheduler_tests.cpp:
```diff
@@ -155,4 +155,45 @@ BOOST_AUTO_TEST_CASE(singlethreadedscheduler_ordered)
+    scheduler.MockForward(boost::chrono::seconds(5*60));
+
+    // ensure scheduler has chance to process all tasks queued for before 1 ms from now.
+    scheduler.scheduleFromNow([&]{ scheduler.stop(false); }, 1);
```
nit, just capture `&scheduler`.

amitiuttarwar replied:
I'm curious- is this for style or is there a difference in behavior? 
from what I can tell the behavior should be the same?
docs: "& (implicitly capture the used automatic variables by reference)" 
compiler explorer: https://godbolt.org/z/QfeAVF

JeremyRubin replied:
I think & is perfectly clear, it's just for style.

I agree in principle if a lambda is multiple lines and has arguments it's usually better to explicitly capture, but I think implicit capture is totally reasonable here.

promag replied:
Yes, should be same, however in L163 you already explicitly capture `&counter` so I though I would like to do the same here.

amitiuttarwar replied:
updated

promag reviewed (commented):
FWIW there's no coverage for `mockscheduler` RPC, just a simple call in the test framework would be nice.

JeremyRubin commented:
@promag I think it's reasonable that the mockscheduler RPC doesn't have coverage presently -- @amitiuttarwar has a larger project that this testing harness was extracted from as a reviewable chunk, a later PR will be using it extensively to test rebroadcasting logic (see https://github.com/bitcoin/bitcoin/pull/18038). I think it's reasonable that the coverage in RPC tests lives there & there isn't a separate one just testing the testing rpc, but feel free to disagree.

amitiuttarwar commented:
thanks for review @promag!

RE testing mockscheduler RPC- as @JeremyRubin mentioned, my next PR uses it [here](https://github.com/bitcoin/bitcoin/pull/18038/files?file-filters%5B%5D=.py#diff-5e6f5e6aab04f82ec4b58148c4bdec72R68) , so the happy path will be tested. But if you prefer separate explicit test coverage, I can add. 

I'll incorporate style nits if I push & invalidate current ACKs

sipa commented on src/rpc/misc.cpp:
```diff
@@ -366,6 +368,34 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+            RPCExamples{""},
+        }.Check(request);
+
+    if (!Params().MineBlocksOnDemand()) {
```
I realize this is copied from `setmocktime`, but it feels very strange to use MineBlocksOnDemand to gate time mocking. Perhaps it's worth adding a separate chainparams property for time mocking, and then using it for both `setmocktime` and `mockscheduler`.

JeremyRubin replied:
@sipa what would you think about just abstracting out a new function named `AllowMockableTime()` and having it call MineBlocksOnDemand  -- that way someone can clean up the logic later, I'm sure we're doing other weird things like this that could also be fixed.

JeremyRubin replied:
Or just making it the responsibility of follow up work, I don't think it's that bad (if someone reviewed using MineBlocksOnDemand for this once... it should still be fine now)

amitiuttarwar replied:
I agree that using `MineBlocksOnDemand` felt weird. Added `m_is_mockable_chain` to chainparams & implemented for `mockscheduler`. If it looks okay I'll do a follow up PR to use that for `mocktime`.

amitiuttarwar committed:
[test] add chainparams property to indicate chain allows time mocking

amitiuttarwar commented:
changes: 
- `mockscheduler` rpc updated to use new `m_is_mockable_chain` chainparams property instead of `MineBlocksOnDemand`
- added checks to `mockscheduler` and `MockForward` methods to ensure `delta_seconds` is within acceptable range 

other comments: 
-  considered adding functional test to test `mockscheduler` rpc, but testing the success case would require actually invoking something scheduled from the c++ code, making it look extremely similar to the PR that provides context for & builds on this one. [link to test](https://github.com/bitcoin/bitcoin/pull/18038/files?file-filters%5B%5D=.py#diff-5e6f5e6aab04f82ec4b58148c4bdec72R68). So I don't think it makes sense to test independently. 

So I believe all existing review comments are addressed. 

This PR is ready for review :)

amitiuttarwar commented:
s390x travis build is failing with "Disk space is too low!" error. everything else is green 

appveyor `wallet_resendwallettransactions.py` failed, seems unlikely that its related but not impossible since `MaybeResendWalletTxs` is called by the scheduler. investigating.

MarcoFalke commented on src/test/scheduler_tests.cpp:
```diff
@@ -155,4 +155,45 @@ BOOST_AUTO_TEST_CASE(singlethreadedscheduler_ordered)
+    size_t num_tasks = scheduler.getQueueInfo(first, last);
+    BOOST_CHECK(num_tasks == 3);
+
+    boost::thread scheduler_thread(std::bind(&CScheduler::serviceQueue, &scheduler));
```
in commit 3679f096c2ef52fa8673e5cd4064154a1889fa12

No features of boost::thread are needed here, so you might as well use a plain std thread
```suggestion
    std::thread scheduler_thread([&]() { scheduler.serviceQueue(); });
```

amitiuttarwar replied:
addressed in follow up

MarcoFalke commented on src/test/scheduler_tests.cpp:
```diff
@@ -155,4 +155,45 @@ BOOST_AUTO_TEST_CASE(singlethreadedscheduler_ordered)
+
+    // check that the queue only has one job remaining
+    num_tasks = scheduler.getQueueInfo(first, last);
+    BOOST_CHECK(num_tasks == 1);
```
in commit 3679f096c2ef52fa8673e5cd4064154a1889fa12:

```suggestion
    BOOST_CHECK_EQUAL(num_tasks, 1);
```

amitiuttarwar replied:
yup. using `boost_check_equal` was causing a warning I didn't understand, but after pushing I realized I just need to cast to `size_t`.

amitiuttarwar replied:
addressed in follow up

MarcoFalke commented on src/chainparams.cpp:
```diff
@@ -231,7 +232,7 @@ class CTestNetParams : public CChainParams {
         fRequireStandard = false;
         m_is_test_chain = true;
-
+        m_is_mockable_chain = false;
```
In commit 3392d690cfb338e1a9a9d7347710c0744e4486d5:

this should be true, otherwise one has to change this in the source code when wanting to test on testnet. While there is an argument for test params to be as close to main params as possible, I think this does not apply to params that have been introduced for primarily testing purposes. (See also my `m_is_test_chain`, which turned out to be too general and people prefer a per-parameter toggle).

So I think this should be true to not hinder testing on testnet.

amitiuttarwar replied:
hmmm, well if we want `m_is_mockable_chain` to be true on testnet, wouldn't it be better not to introduce the param and use the `m_is_test_chain` instead?

amitiuttarwar replied:
Although, conceptually, I don't know if I'm convinced. Mocking time seems useful for increasing automated test coverage, but running the node on testnet simulating closer to mainnet makes sense. But I feel pretty indifferent since even if time is allowed to be mocked on testnet, a user doesn't have to invoke it. And there's no concern around attack because its testnet. Mostly just trying to move the PR forward.

MarcoFalke replied:
Feel free to ignore, this can be handled in the follow-up

amitiuttarwar replied:
not addressed, but noted as open question in the follow up https://github.com/amitiuttarwar/bitcoin/pull/2#issuecomment-584885826. going to resolve here.

MarcoFalke commented on src/rpc/misc.cpp:
```diff
@@ -366,6 +368,32 @@ static UniValue setmocktime(const JSONRPCRequest& request)
 
+static UniValue mockscheduler(const JSONRPCRequest& request)
+{
+        RPCHelpMan{"mockscheduler",
```
in commit cf04e02bbe14257a1513b029231cb6a21c9766ac:

```suggestion
    RPCHelpMan{"mockscheduler",
```

Excessive whitespace?

amitiuttarwar replied:
hmm, I'm seeing the same amount of whitespace on the other `RPCHelpMan` calls in `rpc/misc` & `rpc/net` & `wallet/rpcwallet`. Are you seeing inconsistency? Is there something I'm missing?

MarcoFalke replied:
Feel free to ignore this, but we haven't re-formatted them in b6fb617aaaad5f9cdd7f2ad2825b253ca792055d because they are still in flux.

amitiuttarwar replied:
addressed in follow up

MarcoFalke commented on src/rpc/misc.cpp:
```diff
@@ -366,6 +368,32 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+        throw std::runtime_error("delta_time must be between 0 and 3600 seconds (1 hr)");
+    }
+
+    g_rpc_node->scheduler->MockForward(boost::chrono::seconds(delta_seconds));
```
in commit cf04e02bbe14257a1513b029231cb6a21c9766ac:

Would be nice to protect against null pointer dereference. See `EnsureMemPool`

amitiuttarwar replied:
addressed in follow up

maflcko reviewed (approved):
ACK cf04e02bbe14257a1513b029231cb6a21c9766ac only some style nits 🔵

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK cf04e02bbe14257a1513b029231cb6a21c9766ac only some style nits 🔵
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgrVAwAzQPXXwQxaHhiRYtQO0W21qQLUyCF92Bvjtq2RMMQnmkrQZ8sIeyFkka9
Mm0FDE2e0eU4en2Xw4CK/7yFsWUxlOMx8vrOlzYttO9htkmVT95Rnm7HOuAShvav
lcuEloyfjMe0bKAR+X/X51/roJhf+WSGrKsEFI8C+FV/bu/AbOOxlLEFiW9v6Wo3
83aqh74lXb5B3Bf4aZJRSTAJhlUE9XNlMH664laqefks2HXyseZpvomRd/w90NX/
VC9UzgeTUnSqUAGd278nFFAf1J5tqIfnE37USY1EokaZw4Yc8zsdYOYHUK1ZklEw
ZHIF3WPRln9OaR7kLfWtHslQUnCo1797qP2ouWAVpgRVXl63cNmjMqRGJErGZtXK
23bKPE3D1CgEdb+dO422dKlbiI4Nsdg0pkwnAR/z+KLBwCu+oov2KF3kPhwVdQfq
pUNQMn/cS7X9QP7bHxQOlAnpl4ijuBJfbV+wqU+0fY/WfclKyM/20u7IFecoZFOz
CYu4fn+Z
=NQgO
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6c8ec4d2d3fd87097e17c0ce53d45761554026de152c054c527f36f2fdc7f39b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086c8ec4d2d3fd87097e17c0ce53d45761554026de152c054c527f36f2fdc7f39bf010e9258021b32dc790decdd85db40a0de308fff0108d723067c0f56da767349d5d37f385ee08f1045e4085e6f008c5e8af49e6945c970083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010ade97b5cd2d8fa806ef69c9617fc990808f1045e4085e6f008e4d9ac9ffaafd7260083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010ed6ad0768556d676375b9ff70ae502e708f1045e4085e6f0088897d71bcc277d370083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010fbd8062f86ffb8ad91b997b2952afd1308f1045e4085e6f008ca836d2db3fae5910083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on src/init.cpp:
```diff
@@ -1269,6 +1269,9 @@ bool AppInitMain(NodeContext& node)
     threadGroup.create_thread(std::bind(&TraceThread<CScheduler::Function>, "scheduler", serviceLoop));
 
+    assert(!node.scheduler);
+    node.scheduler = &scheduler;
```
The same should be done in `src/test/util/setup_common.cpp`, no?

amitiuttarwar replied:
addressed in follow up

amitiuttarwar commented:
Thank you for the review @MarcoFalke ! I addressed your concerns in a commit (with one open question), currently opened as a PR on my fork: https://github.com/amitiuttarwar/bitcoin/pull/2, will make as a follow up PR if this one gets accepted. 

Tried to investigate the appveyor failure and cannot see how it could possibly be related to my changes.

amitiuttarwar commented on src/scheduler.h:
```diff
@@ -55,6 +55,11 @@ class CScheduler
     // need more accurate scheduling, don't use this method.
     void scheduleEvery(Function f, int64_t deltaMilliSeconds);
 
+    // Regtest only- mock the scheduler to fast forward in time
```
addressed in follow up

amitiuttarwar commented on src/scheduler.cpp:
```diff
@@ -114,6 +114,28 @@ void CScheduler::scheduleFromNow(CScheduler::Function f, int64_t deltaMilliSecon
 
+void CScheduler::MockForward(boost::chrono::seconds delta_seconds)
+{
+    assert(delta_seconds.count() > 0 && delta_seconds.count() < 3600);
```
addressed in follow up

MarcoFalke commented:
I'd prefer to fixup the style fixes you want to make into the commits in this pull or not do them at all. Splitting them into a separate pull only causes noise in the git history and on GitHub.

amitiuttarwar committed:
[lib] add scheduler to node context

- also update test setup & access point in denial of service test

amitiuttarwar committed:
[rpc] expose ability to mock scheduler via the rpc

amitiuttarwar commented:
incorporated style updates & made `NodeContext.scheduler` a `unique_ptr` instead of raw pointer.

MarcoFalke commented:
Looks like travis fails with the boost exception "invalid argument" (https://github.com/bitcoin/bitcoin/pull/18037#issuecomment-580650220)


```
Running 3 test cases...

Test cases order is shuffled using seed: 902768437

Entering test module "Bitcoin Core Test Suite"

test/scheduler_tests.cpp(11): Entering test suite "scheduler_tests"

test/scheduler_tests.cpp(112): Entering test case "singlethreadedscheduler_ordered"

test/scheduler_tests.cpp(112): Leaving test case "singlethreadedscheduler_ordered"; testing time: 6488us

test/scheduler_tests.cpp(38): Entering test case "manythreads"

test/scheduler_tests.cpp(38): Leaving test case "manythreads"; testing time: 11938us

test/scheduler_tests.cpp(156): Entering test case "mockforward"

terminate called after throwing an instance of 'boost::wrapexcept<boost::condition_error>'

  what():  boost::condition_variable::do_wait_until failed in pthread_cond_timedwait: Invalid argument

unknown location(0): fatal error: in "scheduler_tests/mockforward": signal: SIGABRT (application abort requested)

test/scheduler_tests.cpp(172): last checkpoint

test/scheduler_tests.cpp(156): Leaving test case "mockforward"; testing time: 455us

test/scheduler_tests.cpp(11): Leaving test suite "scheduler_tests"; testing time: 18977us

Leaving test module "Bitcoin Core Test Suite"; testing time: 19091us

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"

amitiuttarwar commented:
yup :( I'm investigating.

MarcoFalke commented on src/scheduler.h:
```diff
@@ -55,6 +55,11 @@ class CScheduler
     // need more accurate scheduling, don't use this method.
     void scheduleEvery(Function f, int64_t deltaMilliSeconds);
 
+    // Mock the scheduler to fast forward in time.
```
style-nit in commit 3f5cebe47b:

Using `/**` will make this a doxygen comment. I know that the other functions don't use doxygen comment, but for new code we should aim for that.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1265,16 +1264,19 @@ bool AppInitMain(NodeContext& node)
+
     // Start the lightweight task scheduler thread
-    CScheduler::Function serviceLoop = std::bind(&CScheduler::serviceQueue, &scheduler);
+    CScheduler::Function serviceLoop = std::bind(&CScheduler::serviceQueue, node.scheduler.get());
```
style-nit in commit 047c1fd23ab7f281af7acd238f3bbb441817686c:

As you are changing this line anyway, might as well get rid of the std::bind mess


```cpp
CScheduler::Function serviceLoop = [&] { node.scheduler->serviceQueue(); };

MarcoFalke commented on src/init.cpp:
```diff
@@ -157,7 +157,6 @@ NODISCARD static bool CreatePidFile()
-static CScheduler scheduler;
 
 void Interrupt(NodeContext& node)
 {
```
style-nit in commit 047c1fd:


It should be safe and a bit cleaner to reset the scheduler unique_ptr at the end of `void Shutdown(NodeContext& node)`.

MarcoFalke commented on src/test/util/setup_common.cpp:
```diff
@@ -103,10 +103,12 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
     g_rpc_node = &m_node;
     RegisterAllCoreRPCCommands(tableRPC);
 
+    g_rpc_node->scheduler = MakeUnique<CScheduler>();
```
style-nit in commit 047c1fd23ab7f281af7acd238f3bbb441817686c:

`g_rpc_node` might go away (#17548), and using `m_node` directly would be cleaner code anyway.

MarcoFalke replied:
Also, the pointer should probably be reset in `~TestingSetup`

amitiuttarwar replied:
yeah totally. do you think the lack of reset could have somehow caused the failure [here](https://travis-ci.org/bitcoin/bitcoin/jobs/649620800)? I'll push shortly and observe, but still trying to fill in my conceptual understanding.

not directly related, but another thing I'm wondering is why the scheduler thread isn't interrupted [here](https://github.com/bitcoin/bitcoin/blob/master/src/init.cpp#L162)

MarcoFalke commented on src/test/util/setup_common.cpp:
```diff
@@ -103,10 +103,12 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
     // from blocking due to queue overrun.
-    threadGroup.create_thread(std::bind(&CScheduler::serviceQueue, &scheduler));
-    GetMainSignals().RegisterBackgroundSignalScheduler(scheduler);
+    threadGroup.create_thread(std::bind(&CScheduler::serviceQueue, g_rpc_node->scheduler.get()));
```
style-nit in commit 047c1fd23ab7f281af7acd238f3bbb441817686c:

Same here

```suggestion
    threadGroup.create_thread([&] { m_node.scheduler->serviceQueue(); });
```

maflcko reviewed (approved):
ACK db45f73ab93e207f792fb5bb11ca1ec2885d8a91 , only some style changes 🖊

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK db45f73ab93e207f792fb5bb11ca1ec2885d8a91 , only some style changes 🖊
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYEACgkQzit1aX5p
pUhXZwwAnlRWgEGarDT3ZkLJNksVg2V3EkG8VEMRyDGnmkUYCRNaeemikKnzxvI7
rrqKckkaYGyZvYqLhgodn+OG+EoZOh3DBrRljH1jdl9DjsNp5VcNVoVV2SHqsdrp
qfoJeMwuzCqo9XYluZ5pKvJCyXYu+JbKIkhojeLRO4G8G6xfR6vejRDRXeG8wnEL
KBkFHveCBBsem3JXwjTgj2uxRLzArVg0x2/shMDMpCsDyWdTn+6ngrTWulAyt5qB
QjlwhCvJ7jyiAp8ymQU0aiRIg7OERYvNbNLxtc3S/DKmMeUq9Tyb1xCQW+y3iD/j
x3kQi0jSief82PhqT4FhdjZPDhu2PbZlVg/M4zghsUiLOA6B/s1M2g8qaqC7OjrW
ucC9PFUen/W0eNl7IIE/SBQ8IfYyDpw/Kz32o7fv/dwUlYUM8qXi6KxYIpplS1OG
pn/Zzh3vdyC6cqPJS0q67+0Crv5at3ac3xx8dF5dTuT5+LHcztXyhANVpJl/y7xt
oXFK3lOi
=G9JB
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1ac115fc816927e857fea47e64737b58665a1508da322004b7d0622a653ebbbf  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081ac115fc816927e857fea47e64737b58665a1508da322004b7d0622a653ebbbff01016c6ac6f4c42719624da6a9b8defa7c408fff0103967f0bf08732d2d28716de97dafc9ee08f020d28da53b02c2fbbf0fbd3ddb83f3c10f7e4084340e8b48893a7f7cb4792e35ca08f1045e457deaf008b21094565fa445470083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01046fe61e4a146b91d0a087f1118c402cd08f120d8e9693f75ddc03fadbaa74b3da1e9393a2021ab13e8b40b965c2d7f1f6029b508f1045e457deaf008ace63e093d12a5ce0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010823b07704d846fa42e7df0e08344df1c08f0203addcea61994c4ef0e491120f0f54b5578b1d5920e433387dd4d56d378d3ab4c08f1045e457deaf008c371dbf77a9e65e50083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f74f7a3c63d15df96dfe857e5dc8f74408f020b5fc4c608a6f5fc675fabdb43f12ecc860ed9795aa93edcc2cfc15f79617280608f1045e457decf008ff82f09eeea7b23b0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
ACK 8bca30ea17cd4c1dacee28eaa27e5fa3493b021d, only change is some style fixups 🕓

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 8bca30ea17cd4c1dacee28eaa27e5fa3493b021d, only change is some style fixups 🕓
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjeVAv/YuSt6tTjGv6SeqSUSmrURxHQKmaO8WtGS57M8zxKiQLp2IlBrXaccECP
ktwXi7z+OxVlfaw/acuEoFZn5kIUjOC19D7tIegGh6qIWJxVYcT2cDghEG3JfTFD
H+gDfpx6+JUjJcXybDwrc0jeOpva0aR6WOV/BBv+JuaOM0G2SXZPFS+DeUkSWGez
8gMhSY4FE1sjGxtYSyhdtlHpROL1qX22joa8kdgVplJIuQJOmn8TGjI/4s4Xu6j7
5Afa60Iqnbqb9paSMyEEi7XPhjOPKtLYWwB/K+itYjvbipaz4uw4yd5+YfMdYy45
F/KSW9+7bYCSGE+ggEUQykrR60CYVPJay5WbwfDs0UdZhKmfyjlbx/iysTx63uKs
4PKeLZoNCJynQA+nodf1vrXuA8nxJgfYdhDP8bWnUCMdcP84FKPdrG7r3v2x6fBt
MeCCRSOXPNEAFtu59xnZw9mc+Dctx5EUyznhy5ybp/3QCUlM9DhxvTbx0H7sFzDo
BTRQiIFP
=2pAG
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `0370e988925c0bd1dc04b6ff2d771181e26d2b0eb3b28b69eb3c0a0976243393  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401080370e988925c0bd1dc04b6ff2d771181e26d2b0eb3b28b69eb3c0a0976243393f01049574cb74c254b4cff2d1668f70cda6108fff01026c290b8151fb0781b6fc0cb81208ee708f1045e4b36fff0088fd6d203aa14532e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0103be0ca71667e5d7ea20c51864f1c399f08f1045e4b36fff0086f80792bcdea6ad20083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01047d88e9bd9e0a63096af6fd0ba3e7d1a08f1045e4b36fff008a56702a3a200defa0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010c2d3773832b121a1b02e763d79bdb94b08f1045e4b36fff0089da1ba29759a55190083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jonasschnelli commented:
This PR seems to have made bitcoinbuilds.org pretty unhappy:

```
  what():  boost::condition_variable::do_wait_until failed in pthread_cond_timedwait: Invalid argument
unknown location(0): fatal error: in "scheduler_tests/mockforward": signal: SIGABRT (application abort requested)
test/scheduler_tests.cpp(172): last checkpoint
```
Example build:
https://bitcoinbuilds.org/index.php?job=5c006a9e-b8e6-4932-a6a8-3ead3d1fc188

Any ideas why we see a SIGABRT?

I think it happens at `std::thread scheduler_thread([&]() { scheduler.serviceQueue(); });` in the scheduler tests.

MarcoFalke commented:
@jonasschnelli See #18037

jonasschnelli commented:
@MarcoFalke: I guess you where trying to refer to #18174. Looking into that now...

jonasschnelli commented:
Master (`Linux 64 depends tests`) fails when using ccache cache (seems to be reproducible 100%):
https://bitcoinbuilds.org/index.php?job=c1d3a0f5-7d47-40e6-9c9b-cac397c5ff30

Master succeeds when clearing the ccache cache (but keeping the dependency cache):
https://bitcoinbuilds.org/index.php?job=f28f0356-d431-4428-832b-49bc870042f5
